### ISSUE TITLE
Extract Agent plugin interface repository

### ISSUE THREAD

**Issue by mssalvatore** [guardicore/monkey#3922]
## Goal

Extract classes, constants, and functions that make up the agent plugin interface.

## Motivation
In order for 3rd-party developers to develop their own plugins efficiently, plugins' dependencies should be installable with `pip`. Extracting the agent/plugin interface into its own repository as we did with [monkeytypes](https://github.com/guardicode/monkeytypes), #3756 helps enable this.

See also #3725.

## Details
- Complete https://github.com/guardicore/monkey/issues/3751 before starting this issue
- A "best effort" approach will be taken to preserve the git history of the files that are moved. See monkeytypes for an example
- The script and artifacts used to migrate monkeytypes is linked below for reference
- Boilerplate for setting up the new project can be created from the [infection monkey cookiecutter template](https://github.com/guardicode/infection-monkey-cookiecutter).
- [cruft](https://pypi.org/project/cruft) should be used to apply the cookiecutter to the migrated project
- The new package will be published to PyPI so it can be installed via `pip`. @mssalvatore can help.
- An effort has been made to identify the components that should be moved to this new repository. The "AgentPluginSpec" box in [this diagram](https://drive.google.com/file/d/1KTrOcWd8jPrmOLu_3mkeGPAfx-OTjfyc/view?usp=sharing) contains that information, but may be incomplete or outdated.

## Resources
- https://github.com/guardicore/monkey/issues/3798
- https://drive.google.com/file/d/1KTrOcWd8jPrmOLu_3mkeGPAfx-OTjfyc/view?usp=sharing
- https://github.com/guardicore/monkey/files/12929661/extract-monkeytypes.zip

**Comment by ilija-lazoroski** [comment]
[extract-monkeyagentplugininterface.zip](https://github.com/guardicore/monkey/files/13812100/extract-monkeyagentplugininterface.zip)

This brings us to a somewhat working repo with passing UTs. We decided to abandon this as it depends on Utils.

**Comment by mssalvatore** [comment]
> [extract-monkeyagentplugininterface.zip](https://github.com/guardicore/monkey/files/13812100/extract-monkeyagentplugininterface.zip)
> 
> This brings us to a somewhat working repo with passing UTs. We decided to abandon for now this as it depends on Utils.

Not abandon, just postpone.

**Comment by ilija-lazoroski** [comment]
Yep, poor choice of words.

**Comment by shreyamalviya** [comment]
> > [extract-monkeyagentplugininterface.zip](https://github.com/guardicore/monkey/files/13812100/extract-monkeyagentplugininterface.zip)
> > This brings us to a somewhat working repo with passing UTs. We decided to abandon for now this as it depends on Utils.
> 
> Not abandon, just postpone.

Only until we extract the plugin utils repository?

**Comment by mssalvatore** [comment]
Where sensible, this package should contain interfaces, not concrete implementations. For example, we may want to write an `ITCPPortSelector` interface for this package and leave the `TCPPortSelector` implementation in the agent.

**Comment by ilija-lazoroski** [comment]
> > > [extract-monkeyagentplugininterface.zip](https://github.com/guardicore/monkey/files/13812100/extract-monkeyagentplugininterface.zip)
> > > This brings us to a somewhat working repo with passing UTs. We decided to abandon for now this as it depends on Utils.
> > 
> > 
> > Not abandon, just postpone.
> 
> Only until we extract the plugin utils repository?

Only until we do #3984

**Comment by ilija-lazoroski** [comment]
> Where sensible, this package should contain interfaces, not concrete implementations. For example, we may want to write an `ITCPPortSelector` interface for this package and leave the `TCPPortSelector` implementation in the agent.

We need to do that in this issue?

**Comment by mssalvatore** [comment]
> > Where sensible, this package should contain interfaces, not concrete implementations. For example, we may want to write an `ITCPPortSelector` interface for this package and leave the `TCPPortSelector` implementation in the agent.
> 
> We need to do that in this issue?

Might as well. Otherwise the agent/plugin interface would be improperly defined.

**Comment by mssalvatore** [comment]
The new repository is at https://github.com/guardicorde/agentpluginapi. The Package is at https://pypi.org/project/monkey-agentpluginapi/
